ls1intum/Artemis
# [Feature] anonymize submissions

### Checklist

### Motivation and Context
Currently a students submission repository still contains lots of data about the student, preventing real blind assessments. In the long run reviewers should not have access to the submission repositories but instead only receive an anonymized version. This PR aims to provide such a version, although currently optionally. It will later be mandatory.

### Description
Adds a new option to the "download submission" option in an assessment that anonymizes the submission before downloading. Tutors don't get to choose but always have it enabled. In detail it:
 - overwrites all author/committer data from all commits
 - deletes all remotes and remote tracking branches
 - deletes all reflogs

Limitations:
 - Only anonymizes the main branch
 - Non-linear histories might lead to unexpected behavior

### Steps for Testing

1. Log in to Artemis
2. Create/Locate a programming exercise
3. Create a submission with multiple commits
4. Assess the submission as tutor through the assessment dashboard
5. Download the submission (as zip)
6. It should not be possible to retrieve the name of the student from this repository by any means (e.g. git log, git log --all, git log --reflog, git branch -a, git remote -v, viewing git blame, ...)
7. The history should still be viewable
8. Assess as instructor
9. Instructors should be able to choose the option when downloading
10. Not anonymizing should result in the unaltered repository
11. Any combination of squash and anonymize should work

### Test Coverage
* GitService.java: 76% -> 82%
* ProgrammingExerciseExportService: 67% -> 68%
* Client unchanged
